Assets/Tests/Runtime/NetworkedMessageChannelTests.cs

```diff
@@ -0,0 +1,204 @@
+using Unity.Netcode;
+using UnityEngine;
+using Unity.Netcode.TestHelpers.Runtime;
+using UnityEngine.Assertions;
+using UnityEngine.TestTools;
+
+namespace Unity.Multiplayer.Samples.BossRoom.Tests.Runtime
+{
+    public class NetworkedMessageChannelTests : NetcodeIntegrationTest
+    {
+        struct EmptyMessage : INetworkSerializeByMemcpy { }
+
+        struct GenericMessage : INetworkSerializeByMemcpy
+        {
+            public bool value;
+        }
+
+        protected override int NumberOfClients => 2;
+
+        NetworkManager FirstClient => m_ClientNetworkManagers[0];
+        NetworkManager SecondClient => m_ClientNetworkManagers[1];
+
+        DisposableGroup m_Subscriptions;
+
+        DIScope m_FirstClientScope;
+        DIScope m_SecondClientScope;
+        DIScope m_ServerScope;
+
+        protected override void OnServerAndClientsCreated()
+        {
+            m_FirstClientScope = new DIScope();
+            m_FirstClientScope.BindInstanceAsSingle(FirstClient);
+            m_FirstClientScope.FinalizeScopeConstruction();
+
+            m_SecondClientScope = new DIScope();
+            m_SecondClientScope.BindInstanceAsSingle(SecondClient);
+            m_SecondClientScope.FinalizeScopeConstruction();
+
+            m_ServerScope = new DIScope();
+            m_ServerScope.BindInstanceAsSingle(m_ServerNetworkManager);
+            m_ServerScope.FinalizeScopeConstruction();
+
+            base.OnServerAndClientsCreated();
+        }
+
+        protected override IEnumerator OnTearDown()
+        {
+            m_Subscriptions.Dispose();
+            m_FirstClientScope.Dispose();
+            m_SecondClientScope.Dispose();
+            m_ServerScope.Dispose();
+            return base.OnTearDown();
+        }
+
+
+        [UnityTest]
+        public IEnumerator EmptyNetworkedMessageIsReceivedByAllSubscribersOnMultipleClients()
+        {
+            var emptyMessageChannelClient1 = new NetworkedMessageChannel<EmptyMessage>();
+            m_FirstClientScope.InjectIn(emptyMessageChannelClient1);
+            var emptyMessageChannelClient2 = new NetworkedMessageChannel<EmptyMessage>();
+            m_SecondClientScope.InjectIn(emptyMessageChannelClient2);
+            var emptyMessageChannelServer = new NetworkedMessageChannel<EmptyMessage>();
+            m_ServerScope.InjectIn(emptyMessageChannelServer);
+
+            var nbMessagesReceived = 0;
+
+            m_Subscriptions = new DisposableGroup();
+            m_Subscriptions.Add(emptyMessageChannelClient1.Subscribe(OnEmptyMessageReceived));
+            m_Subscriptions.Add(emptyMessageChannelClient2.Subscribe(OnEmptyMessageReceived));
+
+            void OnEmptyMessageReceived(EmptyMessage message)
+            {
+                nbMessagesReceived++;
+            }
+
+            emptyMessageChannelServer.Publish(new EmptyMessage());
+
+            // wait for the custom named message to be sent on the server and received on the clients
+            yield return null;
```

Comment:
There should be a "WaitForCondition" in those test utilities somewhere, you can use this in all your tests

Reply:
do you mean like creating a custom yield instruction that waits for two frames so that we can reuse it in all the tests?

Reply:
no, I mean wait for the condition mentioned in your comment

> wait for the custom named message to be sent on the server and received on the clients

AFAIK you're not validating in your test that it takes 2 frames to receive that message, you're just validating it's received at all. So to make sure your test doesn't fail if that behaviour changes, you can yield return new WaitForCondition

Reply:
But the condition is the thing I'm testing for. I'm not sure I understand how that would work here. With a `new WaitUntil(m_NbMessagesReceived==expected)` for example, it would hang indefinitely if the client doesn't receive the expected number of messages and would return early with a success if the client receives more than the expected number of messages

Reply:
Can you guarantee it should always be two frames of wait time?
you can have timeouts in those wait conditions if it never reaches it or goes beyond.

Reply:
I cannot guarantee that. Timeouts do prevent indefinitely hanging, but not early returns with successes. For example, some of these tests expect 0 messages to be received, but I still want to make sure to wait enough time to detect if more messages are received so that the test can fail.

To that end, I ended up changing waiting for two frames to waiting for 0.2 seconds, as is the case in NGO's custom named messages tests.

Reply:
Waiting for times has issues as well. you don't have a guarantee you'll have more than a frame during that time. If there's a hang, that frame could last 0.2+ seconds.

> but I still want to make sure to wait enough time to detect if more messages are received so that the test can fail.

How many of those do you have? Thinking of it, we want our tests to execute as fast as possible. If we have big times like this for a bunch of tests, it'll just increase the time it takes for our PRs to be approved. Are those waits critical enough to add delays to ALL our PRs approvals?

Reply:
> How many of those do you have?

with all the tests, it amounts to a total wait time of 12.8 seconds.

> Waiting for times has issues as well. you don't have a guarantee you'll have more than a frame during that time. If there's a hang, that frame could last 0.2+ seconds.

Should I go back to waiting two frames then? And adjust these tests if the sdk changes this behaviour in the future?